[core] [android] - configurable API endpoint

tobrun opened the pull request:
This PR enables configuration of a custom URL base endpoint. 
Review c++: @jfirebaugh @tmpsantos 
Review java: @zugaldia 

Side notes vs GL-JS implementation:
 -  I went for using APIBaseURL over APIURL for readability 
 -  I initially mimicked the Config class concept but removed it after discussion with @tmpsantos

jfirebaugh commented on platform/default/online_file_source.cpp:
```diff
@@ -152,30 +152,34 @@ OnlineFileSource::~OnlineFileSource() = default;
 std::unique_ptr<AsyncRequest> OnlineFileSource::request(const Resource& resource, Callback callback) {
     Resource res = resource;
+    
+    if(apiBaseURL.empty()){
```
Set the default value to `mbgl::util::API_BASE_URL` instead of checking for the empty string.

jfirebaugh commented on include/mbgl/util/constants.hpp:
```diff
@@ -43,6 +43,8 @@ constexpr Seconds CLOCK_SKEW_RETRY_TIMEOUT { 30 };
 
 constexpr UnitBezier DEFAULT_TRANSITION_EASE = { 0, 0, 0.25, 1 };
 
+constexpr const char* API_BASE_URL = "https://api.mapbox.com/";
```
Let's remove the trailing slash from this URL. That matches GL JS and is more correct -- a trailing slash is a path, and in this case the point is that this is a base URL to which a path will be appended.

jfirebaugh commented on src/mbgl/util/mapbox.cpp:
```diff
@@ -35,19 +34,19 @@ std::vector<std::string> getMapboxURLPathname(const std::string& url) {
     }
-
+    
     return baseURL + "v4/" + url.substr(protocol.length()) + ".json?access_token=" + accessToken + "&secure";
```
After removing the slash from the default value, add a slash here (`"/v4/"`) and in each of the other normalization functions.

tobrun commented:
Thanks @jfirebaugh for 👀 , the remarks have been fixed in ad5f363. 
I will now look into resolving this for the other bindings.

zugaldia commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMapOptions.java:
```diff
@@ -957,6 +978,8 @@ public boolean equals(Object o) {
         if (!Arrays.equals(myLocationBackgroundPadding, options.myLocationBackgroundPadding))
             return false;
         if (style != null ? !style.equals(options.style) : options.style != null) return false;
+        if (apiBaseUrl != null ? !apiBaseUrl.equals(options.apiBaseUrl) : options.apiBaseUrl != null)
```
@tobrun This is very hard to read, could you split it up?

tobrun replied:
This is generated code from intelij or actually it's copied code from intelij. The correct way should be to regenerate the hashcode/equals method.. Will pick this up with other remarks but I'm not taking readability into account with generated code.

zugaldia replied:
@tobrun Sorry, I missed this was generated code. Scratch my comment.

zugaldia commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/http/HTTPRequest.java:
```diff
@@ -56,15 +56,15 @@ private HTTPRequest(long nativePtr, String resourceUrl, String userAgent, String
             HttpUrl httpUrl = HttpUrl.parse(resourceUrl);
             final String host = httpUrl.host().toLowerCase(MapboxConstants.MAPBOX_LOCALE);
-            if (host.equals("mapbox.com") || host.endsWith(".mapbox.com")) {
+            if (host.contains("mapbox.")) {
```
@tobrun Not sure about this one, [from the docs](https://square.github.io/okhttp/3.x/okhttp/okhttp3/HttpUrl.html) the host can be something like `www.google.com` so if a third-party names their server `mapbox.mycompany.com` this will be `true` too.

tobrun replied:
After convo, we need to explicitly set the different supported domains.

tmpsantos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/http/HTTPRequest.java:
```diff
@@ -56,15 +56,15 @@ private HTTPRequest(long nativePtr, String resourceUrl, String userAgent, String
                 resourceUrl = resourceUrl + "events=true";
             }
-
+            
```
nit

tmpsantos commented on platform/android/src/jni.cpp:
```diff
@@ -424,6 +424,13 @@ jni::jobject* nativeGetClasses(JNIEnv *env, jni::jobject* obj, jlong nativeMapVi
     return std_vector_string_to_jobject(env, nativeMapView->getMap().getClasses());
 }
 
+void nativeSetAPIBaseURL(JNIEnv *env, jni::jobject* obj, jlong nativeMapViewPtr, jni::jstring* url){
```
`) {`

tmpsantos commented on platform/default/default_file_source.cpp:
```diff
@@ -28,6 +28,10 @@ class DefaultFileSource::Impl {
         : offlineDatabase(cachePath, maximumCacheSize) {
     }
+    
+    void setAPIBaseURL(const std::string& url){
```
`) {`

tmpsantos commented on platform/default/online_file_source.cpp:
```diff
@@ -152,30 +152,30 @@ OnlineFileSource::~OnlineFileSource() = default;
 std::unique_ptr<AsyncRequest> OnlineFileSource::request(const Resource& resource, Callback callback) {
     Resource res = resource;
-
+    
```
nit

tmpsantos commented on src/mbgl/util/mapbox.cpp:
```diff
@@ -35,19 +34,19 @@ std::vector<std::string> getMapboxURLPathname(const std::string& url) {
     }
-
-    return baseURL + "v4/" + url.substr(protocol.length()) + ".json?access_token=" + accessToken + "&secure";
+    
```
nit

tmpsantos commented:
This PR is missing unit tests for a new core API. ;-)

tobrun commented:
@tmpsantos I fixed up the comments and added a tests in https://github.com/mapbox/mapbox-gl-native/pull/6309/commits/4db1f6d9ab3bf8b84ec6b8baeb139519946ee558.

1ec5 commented:
Tracking the necessary iOS/macOS SDK changes in #6346.

tobrun commented:
Need to fix the tests for glfw, the tests are missing runloop

tobrun commented:
@tmpsantos thanks for the hint on RunLoop. 
This is now implemented in 5825a6c and the tests are passing 😎

tmpsantos commented on test/storage/online_file_source.cpp:
```diff
@@ -409,10 +409,9 @@ TEST(OnlineFileSource, ChangeAPIBaseURL){
     util::RunLoop loop;
     OnlineFileSource fs;
     
-    loop.run();
```
Would be great if you could split the PR in two or three commits:

- One adding the API to core
- One adding the API to android
- One adding the tests

This makes reviewing a lot easier, since reviewing is usually done "per commit". You can do this next time, but for now I'm fine squashing everything into a single commit. The last two are fixing something on the first one, so they don't really add value to the PR.

tmpsantos commented on test/util/mapbox.cpp:
```diff
@@ -10,85 +11,85 @@ using namespace mbgl;
     EXPECT_EQ(
         "https://api.mapbox.com/v4/user.map.json?access_token=key&secure",
-        mbgl::util::mapbox::normalizeSourceURL("mapbox://user.map", "key"));
+              mbgl::util::mapbox::normalizeSourceURL(util::API_BASE_URL, "mapbox://user.map", "key"));
```
Nit, extra tab
